### PR TITLE
CI: Introduce basic OpenVMM testing and utilize reusable actions

### PR DESCRIPTION
**Pull request by maheeraeron**
Adds basic Ubuntu boot testing with OpenVMM from the latest successful main run, and also organizes certain steps into github actions. Furthermore, external github actions are now pinned to a commit SHA.

The OpenVMM tests are limited to X64 for now. ARM64 support will come in the future.

### CODE REVIEW

File: .github/workflows/openvmm-test.yml [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+          echo "run-id=$run_id" >> "$GITHUB_OUTPUT"
+          echo "head-sha=$head_sha" >> "$GITHUB_OUTPUT"
+          echo "Using OpenVMM run $run_id at $head_sha"
+
+      - name: Download OpenVMM X64 test artifacts
+        uses: actions/download-artifact@d3f86a106a0bac45b974a628896c90dbdf5c8093 # v4.3.0
+        with:
+          repository: microsoft/openvmm
+          run-id: ${{ steps.openvmm.outputs.run-id }}
+          github-token: ${{ secrets.OPENVMM_ARTIFACT_TOKEN }}
```
Comment (maheeraeron) [inline .github/workflows/openvmm-test.yml]: need to find out how to get this

Comment (maheeraeron) [reply .github/workflows/openvmm-test.yml]: testing to see if it works without...

File: .github/workflows/platform-ci.yml [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+          target: ${{ matrix.target }}
+          tool-chain: ${{ matrix.tools }}
+
+      - name: Upload build artifacts
+        uses: actions/upload-artifact@ea165f8d65b6e75b540449e92b4886f43607fa02 # v4.6.2
+        with:
+          name: firmware-${{ matrix.target }}-${{ matrix.arch }}-${{ matrix.tools }}-legacy-${{ matrix.c-core }}
+          if-no-files-found: error
+          path: |
+            Build/Msvm${{ matrix.arch }}/${{ matrix.target }}_${{ matrix.tools }}/FV/MSVM.fd
```
Comment (mebersol) [inline .github/workflows/platform-ci.yml:118]: [Copilot] This job will fail for every fork PR, and it gates `release`.

`openvmm-test.yml` declares `OPENVMM_ARTIFACT_TOKEN` as `required: true`, and this passes `secrets: inherit`. Secrets are not available to `pull_request` runs from forks — including this PR, which comes from `maheeraeron/mu_msvm`. Every external contributor's PR will show a red required check they have no way to fix.

Suggest gating the job so it only runs when secrets are actually available:

```yaml
openvmm-test-x64:
  name: OpenVMM Test X64
  needs: build
  if: github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name == github.repository
  uses: ./.github/workflows/openvmm-test.yml
  secrets:
    OPENVMM_ARTIFACT_TOKEN: ${{ secrets.OPENVMM_ARTIFACT_TOKEN }}
```

Two things bundled in there: the `if:` guard, and explicit secret passing instead of `secrets: inherit` (least privilege — the called workflow only needs this one).

Also, can you confirm `OPENVMM_ARTIFACT_TOKEN` is actually provisioned on `microsoft/mu_msvm`? If it isn't yet, this job fails on `main` too.

File: .pytool/CISettings.py [hunk 3]
```diff
@@ -56,6 +57,14 @@ def GetActiveScopes(self):
     def GetDependencies(self):
         return []
 
+    def GetRequiredSubmodules(self):
+        return (
+            RequiredSubmodule("MU_BASECORE"),
+            RequiredSubmodule("Common/MU"),
+            RequiredSubmodule("Feature/DEBUGGER"),
+            RequiredSubmodule("Common/PATINA_EDK2"),
+        )
```
Comment (mebersol) [inline .pytool/CISettings.py:66]: [Copilot] `GetRequiredSubmodules()` and `GetPackagesPath()` now disagree: `Common/PATINA_EDK2` is listed here but the package path just below still returns `(".", "MU_BASECORE", "Common/MU", "Feature/DEBUGGER")`. `MsvmPkg/PlatformBuild.py` lists it in *both*, so CI and platform builds see different package paths.

Either add it to `GetPackagesPath()` for consistency, or add a one-line comment explaining why the CI build doesn't need it on the path.

Separately — the submodule list is now duplicated verbatim in `.pytool/CISettings.py` and `MsvmPkg/PlatformBuild.py`. That's a drift hazard (this PR exists partly because the two drifted). Worth pulling into a shared constant that both import.

File: .github/workflows/openvmm-test.yml [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+          if [[ "${#firmware_files[@]}" -ne 6 ]]; then
+            echo "::error::Expected 6 X64 MSVM.fd files, found ${#firmware_files[@]}"
+            printf '%s\n' "${firmware_files[@]}"
+            exit 1
+          fi
+
+      - name: Resolve latest successful OpenVMM main run
+        id: openvmm
+        shell: bash
```
Comment (mebersol) [inline .github/workflows/openvmm-test.yml]: [Copilot] The guard itself is good, but `6` is an unexplained magic number and the failure message doesn't tell a future maintainer where it comes from.

It's derived from the build matrix: 2 targets x (X64/CLANGPDB legacy TRUE + X64/CLANGPDB legacy FALSE + X64/VS2022 legacy TRUE) = 6. That's non-obvious, and the count silently changes on any intentional matrix edit — e.g. retiring the legacy C core drops it to 4 and turns this red with an error that looks like an artifact-upload failure rather than "you changed the matrix."

At minimum add a comment:

```bash
# 2 targets x 3 X64 toolchain/legacy combos (CLANGPDB TRUE+FALSE, VS2022 TRUE).
# Keep in sync with the build matrix in platform-ci.yml.
if [[ "${#firmware_files[@]}" -ne 6 ]]; then
```

Better still, derive it from the matrix rather than hardcoding.

File: .github/workflows/platform-ci.yml [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            Build/Msvm${{ matrix.arch }}/${{ matrix.target }}_${{ matrix.tools }}/MAP/
+            Build/Msvm${{ matrix.arch }}/${{ matrix.target }}_${{ matrix.tools }}/PDB/
+
+      - name: Upload build logs
+        if: always()
+        uses: actions/upload-artifact@ea165f8d65b6e75b540449e92b4886f43607fa02 # v4.6.2
+        with:
+          name: logs-${{ matrix.target }}-${{ matrix.arch }}-${{ matrix.tools }}-legacy-${{ matrix.c-core }}
+          if-no-files-found: ignore
+          path: |
```
Comment (mebersol) [inline .github/workflows/platform-ci.yml:128]: [Copilot] The explanatory comment that lived on the old `BASE_VERSION` env got dropped in the move:

```yaml
env:
  # Bump this when releasing a new major/minor version. The patch number
  # auto-increments from the latest existing GitHub release with this prefix.
  BASE_VERSION: "26.0"
```

This is the one place a maintainer will look when cutting a new minor version, and without the comment the auto-increment/reset semantics aren't discoverable — you'd have to go read the `jq` expression in `release.yml` to work out what happens to the patch number. Please carry it over:

```yaml
    with:
      # Bump this when releasing a new major/minor version. The patch number
      # auto-increments from the latest existing GitHub release with this prefix.
      base-version: "26.0"
``` #Closed

**Review by mebersol — COMMENTED** [review]
[Copilot] Overall this is a good direction — SHA-pinning the third-party actions, `fail-fast: false`, `if-no-files-found: error`, and the log-upload step are all real improvements, and adding `-legacy-${{ matrix.c-core }}` to the artifact name fixes a latent duplicate-artifact-name collision in the old workflow.

Leaving four comments: one blocking (the fork/secrets issue on `openvmm-test-x64`, which also gates `release`), and three smaller ones around consistency and discoverability.

File: .github/workflows/openvmm-test.yml (hunk 1 above)
Comment (maheeraeron) [reply .github/workflows/openvmm-test.yml]: i think i got past this

File: .github/workflows/platform-ci.yml (hunk 2 above)
Comment (maheeraeron) [reply .github/workflows/platform-ci.yml:118]: This has been addressed. The OpenVMM test no longer requires OPENVMM_ARTIFACT_TOKEN or uses secrets: inherit. It now receives the automatically generated ${{ github.token }} as an explicit composite-action input, with only actions: read and contents: read permissions. This token remains available with restricted read-only permissions on fork pull requests, and the OpenVMM test has passed using this configuration. No repository secret needs to be provisioned.

File: .pytool/CISettings.py (hunk 3 above)
Comment (maheeraeron) [reply .pytool/CISettings.py:66]: Good catch. PATINA_EDK2 is a package root used by the platform DSC, so I’ll add it to CISettings.GetPackagesPath() to keep CI setup consistent with PlatformBuild.py. 

I’m going to defer extracting shared settings for now because these are independent Stuart entry points with different package-path return formats; a shared module would add import coupling for a very small constant list.

File: .github/workflows/openvmm-test.yml (hunk 4 above)
Comment (maheeraeron) [reply .github/workflows/openvmm-test.yml]: Good point. I moved the expected count out of the composite action and made it an explicit expected-firmware-count input at the workflow call site, where it is documented as 2 targets across the three X64 combinations: CLANGPDB legacy true/false and VS2022 legacy true. The action now uses that input and reports that the expected count is based on the build matrix. Fully deriving it would require additional runtime matrix plumbing, so keeping the documented value beside the matrix seemed like the clearer tradeoff.

Comment (maheeraeron) [reply .github/workflows/openvmm-test.yml]: To derive from the matrix itself, may need to defer to a follow up PR that moves how we define our build matrix

Comment (maheeraeron) [reply .github/workflows/openvmm-test.yml]: Actually, i could just change the openvmm action to expect a valid firmware anyway... that's safer

File: .github/workflows/platform-ci.yml (hunk 5 above)
Comment (maheeraeron) [reply .github/workflows/platform-ci.yml:128]: Good catch. I restored the comment beside the base-version input so maintainers can see that changing 26.0 starts a new major/minor release series and that the patch component is calculated automatically from existing releases.

File: .github/workflows/platform-ci.yml [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+        with:
+          persist-credentials: false
+
+      - name: Release firmware
+        uses: ./.github/actions/release
+        with:
+          # Bump this when releasing a new major/minor version. The patch number
+          # auto-increments from the latest existing GitHub release with this prefix.
+          base-version: "26.0"
```
Comment (mebersol) [inline .github/workflows/platform-ci.yml]: >base-version: "26.0" [](http://example.com/codeflow?start=10&length=20)

harder to find down here?

File: MsvmPkg/Docs/BUILDING.md [hunk 7]
```diff
@@ -32,7 +32,7 @@ Run the source checks from the repository root before opening a pull request. In
 tools used by the checks on a new checkout or whenever their configuration changes:
 
 ```powershell
-python -m pip install -r pip_requirement_stable.txt
+python -m pip install -r pip-requirements.txt
```
Comment (mebersol) [inline MsvmPkg/Docs/BUILDING.md:35]: >requirements [](http://example.com/codeflow?start=29&length=12)

what does this change mean? #Closed

Comment (mebersol) [reply MsvmPkg/Docs/BUILDING.md:35]: I see it's just a rename

Comment (maheeraeron) [reply MsvmPkg/Docs/BUILDING.md:35]: this is just to be in line with mu repos, because they also use pip-requirements.txt not `stable`

File: .github/workflows/platform-ci.yml (hunk 6 above)
Comment (maheeraeron) [reply .github/workflows/platform-ci.yml]: ill fix it to be up top

Comment (maheeraeron) [reply .github/workflows/platform-ci.yml]: done

File: .github/workflows/platform-ci.yml [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        uses: actions/setup-python@a26af69be951a213d495a4c3e4e4022e16d87065 # v5.6.0
+        with:
+          python-version: "3.12"
+          cache: pip
+          cache-dependency-path: pip-requirements.txt
+
+      - name: Install Python dependencies
+        shell: pwsh
+        run: python -m pip install -r pip-requirements.txt
+
```
Comment (mebersol) [inline .github/workflows/platform-ci.yml:101]: [Copilot] This `env:` now sits on a step that invokes a **composite action**, whereas previously it was on a `run:` step that called `stuart_build` directly. In the old form the variable was unambiguously in the process environment; env propagation from a calling step into composite sub-steps is much less well-defined.

What makes this worth changing rather than just testing: **the failure mode is silent.** `MsvmPkg/BuildPlugins/FirmwareVersionBlob/FirmwareVersionBlob.py` does

```python
if thebuilder.env.GetValue("OFFICIAL_BUILD", "").strip().lower() not in ("", "0", "false"):
    flags |= self.FW_VERSION_FLAG_OFFICIAL
```

If the variable never arrives, nothing fails — `main` builds just ship with `MSVM_FIRMWARE_VERSION_FLAG_OFFICIAL` clear, and it goes unnoticed until someone inspects a released firmware version record.

The action already declares `architecture`, `c-core`, `target`, and `tool-chain` as inputs. Making this one an input too is unambiguous and consistent:

```yaml
# .github/actions/build-platform/action.yml
inputs:
  official-build:
    description: Set to 1 to mark the build as official in the firmware version record
    required: false
    default: ""
...
    - name: Stuart build
      shell: pwsh
      env:
        OFFICIAL_BUILD: ${{ inputs.official-build }}
      run: >-
        stuart_build
        ...
```

```yaml
# .github/workflows/platform-ci.yml
      - name: Build platform
        uses: ./.github/actions/build-platform
        with:
          architecture: ${{ matrix.arch }}
          c-core: ${{ matrix.c-core }}
          # Mark main-branch builds as official in the embedded firmware version record.
          official-build: ${{ (github.event_name == 'push' && github.ref == 'refs/heads/main') && '1' || '' }}
          target: ${{ matrix.target }}
          tool-chain: ${{ matrix.tools }}
```

Either way, worth confirming on the first `main` build after merge that a released `MSVM.fd` actually has BIT1 set in the version record.

File: .github/actions/openvmm-test/action.yml [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+      run: |
+        response="$(gh api \
+          "repos/microsoft/openvmm/actions/workflows/openvmm-ci.yaml/runs?branch=main&event=push&status=success&per_page=20")"
+
+        run_id=""
+        head_sha=""
+        candidates="$(jq -r '.workflow_runs[] | [.id, .head_sha] | @tsv' <<< "$response")"
+        while IFS=$'\t' read -r candidate_id candidate_sha; do
+          artifact_pages="$(gh api --paginate --slurp \
+            "repos/microsoft/openvmm/actions/runs/$candidate_id/artifacts?per_page=100")"
```
Comment (mebersol) [inline .github/actions/openvmm-test/action.yml:45]: [Copilot] Edge case: if `candidates` comes back empty — an API hiccup, or `openvmm-ci.yaml` gets renamed on the OpenVMM side, which is exactly the kind of thing that eventually happens — `<<< ""` still feeds the loop one empty line.

The body then runs with `candidate_id` unset and calls `repos/microsoft/openvmm/actions/runs//artifacts?per_page=100`, which 404s. Under bash's default `set -e` that kills the step with a raw `gh: Not Found`, instead of the much clearer `::error::No successful OpenVMM main run has all required artifacts` you have waiting a few lines below.

One-line fix:

```bash
        while IFS=$'\t' read -r candidate_id candidate_sha; do
          [[ -z "$candidate_id" ]] && continue
          artifact_pages="$(gh api --paginate --slurp \
            "repos/microsoft/openvmm/actions/runs/$candidate_id/artifacts?per_page=100")"
```

Unrelated to the fix, but worth saying: the whole "scan up to 20 runs and skip ones with expired/missing artifacts" approach is a solid improvement over just taking the newest successful run.

File: .github/workflows/platform-ci.yml [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+  workflow_dispatch:
+
+# Bump this when releasing a new major/minor version. The patch number
+# auto-increments from the latest existing GitHub release with this prefix.
+env:
+  RELEASE_BASE_VERSION: "26.0"
+
+jobs:
+  source-checks:
+    name: Source Checks
```
Comment (mebersol) [inline .github/workflows/platform-ci.yml:19]: [Copilot] `source-checks` no longer sets `core.autocrlf false`, but the `build` job below (line 75) still does. That's backwards: `build` doesn't care about line endings, and `source-checks` is the job that actually runs `LineEndingCheck`.

The old `Platform-Build.yml` had this as the first step of `source-checks`, ahead of the checkout, and that placement was deliberate. The `windows-2022` image ships `core.autocrlf=true` in its system gitconfig, and nothing in the current path overrides it — `actions/checkout` doesn't touch it, and neither does `mu_devops/stuart-ci` (I read the action; it's `stuart_setup` / `stuart_update` / `stuart_ci_build` plus log shuffling, no git config). So files stored with LF in the index get normalized to CRLF in the working tree *before* the check inspects them, and `LineEndingCheck` can pass on files that are LF in the repo.

There is a "Fix yaml line endings" commit in this PR, so something caught something — but I can't tell from here whether that came from CI or from a local run.

Cheap way to settle it: push a deliberately LF-only file and confirm Source Checks goes red. If it doesn't, restore the step:

```yaml
  source-checks:
    name: Source Checks
    runs-on: windows-2022
    steps:
      - name: Disable automatic line ending conversion
        shell: pwsh
        run: git config --global core.autocrlf false

      - name: Enable long paths
        shell: pwsh
        run: git config --system core.longpaths true
```

More durable option: add a `.gitattributes` so the result doesn't depend on runner git config at all.

**Review by mebersol — COMMENTED** [review]
[Copilot] Re-reviewed at `b5b0019`. All four comments from my first pass are properly addressed — dropping the PAT in favour of `${{ github.token }}` is a better answer than the fork guard I suggested, and replacing the magic `-ne 6` with `-eq 0` is the right call given `needs: build` already requires every matrix leg to pass. The run-selection loop that skips OpenVMM runs with expired or missing artifacts, `normalize_bundle_file`, and adding `dependabot.yml` alongside the SHA pinning are all good additions.

Three comments on the new iteration. The `OFFICIAL_BUILD` one is the only one I'd want resolved before merge — silent loss of the official-build flag on `main` is the kind of thing that stays invisible for months.

File: .github/workflows/platform-ci.yml (hunk 8 above)
Comment (maheeraeron) [reply .github/workflows/platform-ci.yml:101]: [Copilot] Addressed in 95b12aaa8d. The composite action now declares an official-build input, and only the Stuart build sub-step maps it explicitly to OFFICIAL_BUILD. YAML parsing, wiring assertions, and editor diagnostics passed.

File: .github/actions/openvmm-test/action.yml (hunk 9 above)
Comment (maheeraeron) [reply .github/actions/openvmm-test/action.yml:45]: [Copilot] Addressed in f1975169b7. The run-selection loop now skips an empty candidate ID, so an empty workflow run response reaches the intended clear error without requesting /runs//artifacts. YAML parsing, Bash syntax, and a synthetic empty-response test passed.

File: .github/workflows/platform-ci.yml (hunk 10 above)
Comment (maheeraeron) [reply .github/workflows/platform-ci.yml:19]: [Copilot] Addressed in 2e23c6ab32. The core.autocrlf=false step now runs before checkout in source-checks, where LineEndingCheck executes, and was removed from the build job. YAML parsing, step-order assertions, and whitespace validation passed.

**Review by mebersol — APPROVED** [review]
